### DISCUSSION TITLE
生成版本信息正确解析的 POM 文件 | 夜天之书

### DISCUSSION THREAD

**Discussion by giscus** [tisonkun/blog#54] [Comments]
# 生成版本信息正确解析的 POM 文件 | 夜天之书

本文成文于 2019 年。最近 Apache StreamPark (Incubating) 项目要做第一个 Apache 版本的发布，遇到了类似的发布多 Scala 支持版本时如何正确生成对应 POM 文件，又尽可能复用流水线的问题。由于过往发布记录都被删除，故重新发布。  近日在阅读 FLINK 代码时发现 FLINK 有一个 force-shading 模块，关于这个模块的作用注释在其使用

https://www.tisonkun.org/2022/12/12/maven-properties-enforce/

**Comment by GOODBOY008** [comment]
你好，Tison 老师，请教一个问题。开源项目的默认分支有什么规范或者约定吗？kafka 是 trunk，flink 是master，rocketmq 是develop ，我们怎么理解这个默认分支？

**Reply by tisonkun** [reply]
没有。

大部分项目是用 Git 来管理，Git 以前默认的主分支叫 `master`，后来因为众所周知的原因换成了 `main`。其他的叫 `trunk`/`dev`/`devel`/`develop`/`unstable` 的都有，没有特别的意义。

**Reply by tisonkun** [reply]
有一些是跟 Git 工作流相关的，对应的工作流作者在编写的时候有自己的偏好。

Anyway, naming is difficult :)

**Reply by GOODBOY008** [reply]
谢谢老师解答~

**Comment by Kontinuation** [comment]
Thank you Tison. This blogpost really saved my day.

https://issues.apache.org/jira/browse/SEDONA-284

**Reply by tisonkun** [reply]
Good to know :D

**Comment by Roiocam** [comment]
Tison 老师，这个方案下似乎也有一个问题那就是不会作用于 POM 模块以及 <parent> 下的 artifactId 。

首先是打包后 <parent>  长这样：

```xml
    <parent>
        <artifactId>parent-${scala.binary.version}</artifactId>
        <groupId>com.example</groupId>
        <version>scala-SNAPSHOT</version>
        <relativePath>../pom.xml</relativePath>
    </parent>
```

其次是对于  ` <packaging>pom</packaging>` 的模块，打包后是这样的

```
    <artifactId>example-dependencies${scala.binary.version}</artifactId>
```

**Reply by Roiocam** [reply]
根据这篇文章的内容，`maven-shade-plugin` 会在打包的时候生成 dependency-reduced-pom.xml，这个就是将 `properties` 具体化的结果，但是由于     <packaging>pom</packaging> 的模块在打包过程中是没有任何产物的，因此 `maven-shade-plugin` 不会作用于此模块。

对于 parent 的  artifactId 为什么在  dependency-reduced-pom.xml  中没有变化，这个就不清楚了

**Reply by Roiocam** [reply]
不对，我知道第二点了，因为我的模块结构是这样的

```
root
----> parent-1 (pom)
----|----> child-1 (jar)
----> parent-2 (jar)
```
因为 `maven-shade-plugin` 不作用于 pom，因此 child-1 拿到的 parent-1 的 artifactId 就是没有改变的结果，而不是 `maven-shade-plugin` 不会转换  <parent> 的 artifactId

**Comment by syhily** [comment]
在网上搜索类似问题时正好搜到此文章，我想说，现在更好的方法是使用 flatten-maven-plugin 来生成对应的 `pom.xml` 文件。
